examples/nlp/language_modeling/megatron_gpt_ckpt_to_nemo.py

```diff
@@ -52,35 +51,74 @@ def get_args():
+        required=False,
+        default=None,
+        help="Custom merge file path; overwrite current path stored in config",
+    )
+
     args = parser.parse_args()
     return args
 
 
-def convert(rank, world_size, args):
+def main():
+    args = get_args()
+    model_config_yaml = "model_config.yaml"
+    model_weights_ckpt = "model_weights.ckpt"
 
-    app_state = AppState()
-    app_state.data_parallel_rank = 0
-    trainer = Trainer(gpus=args.tensor_model_parallel_size)
-    # TODO: reach out to PTL For an API-safe local rank override
-    trainer.accelerator.training_type_plugin._local_rank = rank
+    tmp_dir = tempfile.mkdtemp()
+    t_gpu_num = args.tensor_model_parallel_size
 
-    if args.tensor_model_parallel_size is not None and args.tensor_model_parallel_size > 1:
-        # inject model parallel rank
-        checkpoint_path = os.path.join(args.checkpoint_folder, f'mp_rank_{rank:02d}', args.checkpoint_name)
+    start_time = time.time()
+    config_yaml = os.path.join(tmp_dir, model_config_yaml)
+    if t_gpu_num == 1:
+        sources = [os.path.join(args.checkpoint_folder, args.checkpoint_name)]
+        model_weights = [os.path.join(tmp_dir, model_weights_ckpt)]
     else:
-        checkpoint_path = os.path.join(args.checkpoint_folder, args.checkpoint_name)
-
-    model = MegatronGPTModel.load_from_checkpoint(checkpoint_path, hparams_file=args.hparams_file, trainer=trainer)
-    model._save_restore_connector = NLPSaveRestoreConnector()
-    model.save_to(args.nemo_file_path)
-    logging.info(f'NeMo model saved to: {args.nemo_file_path}')
-
-
-def main() -> None:
-    args = get_args()
-    world_size = args.tensor_model_parallel_size
-    mp.spawn(convert, args=(world_size, args), nprocs=world_size, join=True)
+        sources = [
+            os.path.join(args.checkpoint_folder, f"mp_rank_{i:02d}", args.checkpoint_name) for i in range(t_gpu_num)
+        ]
+        model_weights = [os.path.join(tmp_dir, f"mp_rank_{i:02d}", model_weights_ckpt) for i in range(t_gpu_num)]
+
+    for s, t in zip(sources, model_weights):
+        print("****** Start converting...", s, t)
+        ckpt = torch.load(s, map_location="cpu")
+        conf = ckpt["hyper_parameters"]
+        ckpt = ckpt["state_dict"]
+        os.makedirs(os.path.dirname(t), exist_ok=True)
+        torch.save(ckpt, t)
+
+    if args.vocab_file is not None:
+        conf.tokenizer.vocab_file = args.vocab_file
+        conf.vocab_file = args.vocab_file
+    if args.merge_file is not None:
+        conf.tokenizer.merge_file = args.merge_file
+        conf.merges_file = args.merge_file
+
+    print("****** Conf: ", conf)
+    print("****** Checkpoints processing time: ", time.time() - start_time)
+
+    tar_start_time = time.time()
+    with open(config_yaml, 'w') as fout:
+        OmegaConf.save(config=conf, f=fout, resolve=True)
+
+    print("****** Calling tar...")
+    subprocess.call(f'tar -czvf {args.nemo_file_path} ./*', cwd=tmp_dir, shell=True)
```

Comment:
please use model.save_to(...) API to create .nemo files. Otherwise we are at risk of producing incompatible .nemo files at some point in time.

Reply:
Hey Oleksii, I agree with you. Its safer to call apis. However, I am not familiar with the apis. I thought Eric discussed with Markel. The reason I filed this PR is that Markel told me we could use this simple converter at the moment. If we need to refactor the original code, use barrier or whatever I would probably hand it back to @ericharper.